_posts/2024-07-19-quarkus-and-leyden.adoc

```diff
@@ -27,55 +27,83 @@ From a Quarkus perspective, we've done a fairly good job on all such metrics but
 
 To better understand the scope of the potential improvements, we need to take a step back and discuss how the JVM works today, especially how our application is started and iteratively evolves from interpreting our bytecode to its highest performance mode: running native code which is highly optimized, adapted to the particular hardware, the configuration of the day, and the specific workloads it's been asked to perform. No other runtime is able to match the JVM on this.
 
-By default, as it boots, Java is a lazy interpreted language. The content of our JAR file is not executable machine code, but Java bytecode. The JVM will only compile the bytecode of a class the first time it gets referenced during execution. When a target class cannot be found the program throws a ClassNotFoundException.
+By default, as it boots, Java is a lazy interpreted language. The content of our JAR file is not executable machine code, but Java bytecode. The JVM will only compile the bytecode of a class when it is needed for execution. 
 
```

Comment:
I think this needs reworking to introduce the notion of JIT by stealth and set up for an explanation of how the JVM does it more widely before we go on to contrast with AOT.

First we establish that class laoding is JIT:

"A Java runtime does not directly run Java source code. Instead it consumes class bytecode, generated from Java source code using the javac compiler. A key feature of bytecode is portability, encoding the structure of Java classes and operation of their methods in a machine and operating-system independent format. A Java runtime obeys the type information in the bytecode when laying out Java objects. Execution of a method normally involves interpreting the operations in the method bytecode, although a runtime may also choose to compile method bytecode to equivalent machine code and execute the latter directly.

Most notably, the unit of bytecode is a class file, which models a single class. The Java runtime provides a host of utility and runtime management classes, as class files embedded in either system jars or jmod files. Applications supplement this with their own class files, usually by appending jars to the classpath or module path.

Bytecode is provided class-at-a-time to allow the runtime to load classes /lazily/: i.e. the runtime will only lookup, verify and consume a class file when that class's definition is required to proceed with execution. Lazy loading is what allows Java to be a dynamic language -- i.e. one where the code that is included in the program can be decided at runtime. That might include loading classes from jars selected at runtime or, possibly, loaded from the network. Alternatively, it might include generating class bytecode at runtime, as is done with proxy classes or service provider auxiliary classes."

Now we need to fully dismantle the idea that compilation is what JIT is about:

"Class loading is only one of many operations that a Java runtime performs lazily or 'Just In Time' (JIT). The other well-known example is JIT method compilation, as mentioned above. Indeed, many people assume JIT implies compilation, even though, as we have seen, it also accurately describes Java's lazy loading policy. JIT compiled code can run orders of magnitude faster than the interpreter but compilation consumes time and CPU cycles. So, the cost of compilation is normally only repaid for methods which get executed repeatedly. Java runtimes lazily JIT-compile methods to avoid wasting time generating code that only get run a few times and hence does not repay the compile costs."

Next prepare the idea that JIT has a wider range and contrast it with AOT:

"What is less well known is that many other operations performed by the JVM are done lazily or 'Just In Time'. We'll proceed to explain those things later but the reason for highlighting this wider use of the term is to explain one of optimizations project Leyden is offering. An alternative to doing things 'Just in Time' is to do them 'Ahead Of Time' (AOT). For example, Graal's Native Image runtime loads the bytecode of every single class needed by an application, including JDK runtime classes, 'Ahead Of Time' i.e. at image build time. It uses the type and method information encoded in that bytecode to 'Ahead Of Time' compile a complete program that includes code for every method that might possibly be executed by the application. Graal Native Image lies at one extreme, everything is done AOT, and the traditional Java runtime model lies at the other extreme, everything is done JIT. However, it is actually possible to mix and match AOT and JIT models of execution in one runtime and that is a goal of the latest EA release of project Leyden."

This leaves open the question of whether we are talking about mixing AOT loading with JIT compilation or whether the mixing might mean, say, a mix of AOT and JIT loading. So, to clarify that it is the latter we can now use CDS as an example. This also points out that this is not a new idea and allows us to explain that Leyden is merely taking it further before we go onto say how.

"Indeed, this is not a wholly new idea as far as the OpenJDK runtime is concerned. OpenJDK has supported a mixed AOT/JIT class loading model for years with CDS. The observation that led to CDS being proposed was that most applications load the same classes every time they run, both JDK classes during JDK bootstrap and application classes during application startup and warmup. Loading requires locating a class bytecode file, possibly calling out to a Java ClassLoader, parsing the bytecode then building an JVM-internal model of the class. This internal model unpacks the information packed into the bytecode into a format that enables fast interpreted or compiled execution. If this loading work could be done once and the results efficiently reused then that would save time during startup and warmup.

Initially CDS optimized this process for a large set of core JDK classes,. It worked by running up the JVM and dumping a class model for all classes loaded during startup into an archive file laid out in memory format. The resulting JDK module, class, field and method graph can then be quickly remapped into memory next time the JVM runs. Loading a class that is present in the archive involves a simple lookup in the AOT class model. Loading a class not present in the archive requires the normal JIT steps of bytecode lookup, parsing and unpacking. Subsequent improvements to CDS allowed application classes also to be stored in the CDS archive at the end of a short application training run."

Now we reassure users this is not a wacky idea but something they already benefit from:

"A CDS archive for JDK classes has been built into the JVM from JDK17 onwards, halving JDK startup time. This same mixed model AOT/JIT model provides significant improvements to application startup and warmup times, depending on how well the training run exercises application code. So, selective JIT vs AOT operation is not some new thing. The goal of project Leyden is to address the other JIT operations of the JVM, extending the range of cases where AOT can be selectively used as an alternative to JIT in order to provide greater performance gains."

Now we are in a position where we can start talking about other areas where a JIT vs AOT trade-off is available and explain how Leyden is using AOT, based on the same training run model, to avoid repeated JIT work.

Reply:
I moved the CDS part to the Training and Profiling section, which I think made more sense.

Reply:
I don't believe it helps to postpone the stuff about CDS until after providing details how JIT linking and JIT compilation work. The point of mentioning CDS up front is that it clarifies what it means to do something AOT vs JIT. It uses a specific, concrete JIT operation that is familiar and a relatively simple change that allows it to be done AOT. It also introduces the idea of a 'training run' (actually using that term) and its use to improve a production run with the simplest possible example. This should be (to some degree, at least) familiar and understandable to most Java programmers with a relatively understandable performance trade-off. So, if a reader were to stop -- or start getting confused -- at this point they at least have the understanding that Leyden is playing with shifting runtime work to ahead of time work to avoid having to repeat doing the same laborious work every time the JVM runs.

Explaining the rest of the stuff the JVM does lazily (aot) and what Leyden does to do allow it to be done once (AOT) in a training run and then re-used is better postponed until after the CDS example. That way the basic idea is got across and the later explanation simply adds complexity to a know model. If readers start to drop out from understanding all the details they still get the drift.

Reply:
Also, as part of presenting what Leyden does in these two separate improvement you can (and should) underline the point, as was done for CDS, that the training run enables some of the loading/linking/compilation to be done AOT but that anything not trained for will still be done in the normal JIT way i.e. this is not 'one size fits all' but 'mix and match'.

Reply:
Reordering done. I had to rewrite some parts to make it make sense (as some concepts where now explained after we originally  used them).